presto-main/src/main/java/io/prestosql/sql/planner/DynamicFilter.java

```diff
@@ -0,0 +1,244 @@
+        private final PageChannelSelector selector;
+
+        public Collector()
+        {
+            this.pageBuilder = new ImmutableList.Builder<>();
+            int[] buildChannels = descriptors.stream().mapToInt(d -> d.buildChannel).toArray();
+            this.selector = new PageChannelSelector(buildChannels);
+        }
+
+        public void add(Page page)
+        {
+            // Collect only the columns which are relevant for the JOIN.
+            pageBuilder.add(selector.apply(page));
+        }
+
+        public TupleDomain<ColumnHandle> collect()
+        {
+            List<Page> pages = pageBuilder.build();
+            ImmutableMap.Builder<ColumnHandle, Domain> domainsBuilder = new ImmutableMap.Builder<>();
+            int channel = 0;
+            for (DynamicFilter.Descriptor descriptor : descriptors) {
+                domainsBuilder.put(descriptor.probeColumn, getDomain(pages, channel, descriptor));
+                channel++;
+            }
+
+            return TupleDomain.withColumnDomains(domainsBuilder.build());
+        }
+
+        private Domain getDomain(List<Page> buildPages, int channel, Descriptor descriptor)
+        {
+            ImmutableList.Builder<Object> values = new ImmutableList.Builder<>();
+            boolean nullAllowed = false;
+            for (Page page : buildPages) {
+                Block block = page.getBlock(channel);
+                for (int i = 0; i < block.getPositionCount(); i++) {
+                    Object value = TypeUtils.readNativeValue(descriptor.buildType, block, i);
+                    if (value == null) {
+                        nullAllowed = true;
+                    }
+                    else {
+                        values.add(value);
+                    }
+                }
+            }
+            return Domain.create(ValueSet.copyOf(descriptor.buildType, values.build()), nullAllowed);
+        }
+    }
+
+    public static class Descriptor
+    {
+        private final Type buildType;
+        private final int buildChannel;
+        private final ColumnHandle probeColumn;
+
+        public Descriptor(Type buildType, int buildChannel, ColumnHandle probeColumn)
+        {
+            this.buildType = buildType;
+            this.buildChannel = buildChannel;
+            this.probeColumn = probeColumn;
+        }
+    }
+
+    private final List<Descriptor> descriptors;
+    private final ImmutableList.Builder<CompletableFuture<TupleDomain<ColumnHandle>>> partitions;
+    private final CompletableFuture<CompletableFuture<Optional<TupleDomain<ColumnHandle>>>> result;
+
+    public DynamicFilter(List<Descriptor> descriptors)
+    {
+        this.descriptors = descriptors;
+        this.partitions = new ImmutableList.Builder<>();
+        this.result = new CompletableFuture<>();
+    }
+
+    /// Decorate existing page source provider, to allow dynamic filtering.
+    public PageSourceProvider wrap(PageSourceProvider pageSourceProvider)
```

Comment:
I think it would be cleaner (instead of `PageSourceProvider` wrapper) if `DynamicFilter` provided `ListenableFuture` that provides aggregated `TupleDomain`. Then one could add listener that would `applyDynamicFilter` to `pageSourceProvider`.

Reply:
Done at 39a2eb47da (but I still had to create a "dynamic" page source provider to poll for dynamic filters during the execution of the operator over the table splits - see the `createDynamicFilteringPageSourceProvider` method).

Reply:
> but I still had to create a "dynamic" page source provider to poll for dynamic filters during the execution of the operator over the table splits - see the createDynamicFilteringPageSourceProvider method

Why so? You have access to `PageSource` in `LocalExecutionPlanner`. Why not call `applyDynamicFilters` to it when `TupleDomain` future gets done? I think `ConnectorPageSource` is guaranteed to be thread safe.

Reply:
IIUC, the issue is that the new `PageSourceProvider#applyDynamicFilter` API operates on a `TableHandle`, returning potentially new and filtered `TableHandle` - similarly to how it's done be `Metadata#applyFilter` API.
https://github.com/prestosql/presto/pull/931/files#diff-3bb2ddad9db7856c6a9ce4a76dcd81aaR29
An example implementation for `MemoryPageSourceProvider`:
https://github.com/prestosql/presto/pull/931/files#diff-8b4a37d650086d97552712b506dfb6a8R85

Initially, I have intended that `PageSourceProvider#applyDynamicFilter` would be called just before creating a new `PageSource` for each split with the current dynamic filter (if available) - "restricting" the effective predicate on the `TableHandle` to be used for the current split.
I did it by "wrapping" the existing `PageSourceProvider` to keep the required changes to the `LocalExecutionPlanner` module.

I am not sure that I understand your suggestion above, so please let me know if we need to change some of the existing APIs to make the overall design better.

Reply:
Ok. That makes sense. Thanks!